### ISSUE TITLE
request: hunt group with external number

### ISSUE THREAD

**Issue by manfer** [irontec/ivozprovider#721]
I have done a test editing these lines:

https://github.com/irontec/ivozprovider/blob/e912a9a3b681f5f2f4d41a33fb50e7236496766c/asterisk/agi/src/Agi/Action/HuntGroupAction.php#L110-L114

into:

```
        // Configure the list of users to be called
        if ($huntGroup->getStrategy() == HuntGroupAction::RingAll) {
            $this->agi->setVariable("HG_ID", $huntGroup->getId());
            $this->agi->setVariable("HG_ENDPOINTLIST", join($huntGroupEndpoints,'&'));
            $this->agi->setVariable("HG_TIMEOUTLIST", $huntGroup->getRingAllTimeout());

            // use some external numbers instead
            $this->agi->setVariable("HG_ENDPOINTLIST", "PJSIP/+34xxxxxxxxx@proxytrunks&PJSIP/+34yyyyyyyyy@proxytrunks&PJSIP/+34zzzzzzzzz@proxytrunks");
```
where xxxxxxxxx, yyyyyyyyy, zzzzzzzzz are three mobile numbers.

And when I route a DDI to a hunt group with ring all strategy the system correctly calls all those three mobile numbers and works properly.

And tried too:

```
        // Configure the list of users to be called
        if ($huntGroup->getStrategy() == HuntGroupAction::RingAll) {
            $this->agi->setVariable("HG_ID", $huntGroup->getId());
            $this->agi->setVariable("HG_ENDPOINTLIST", join($huntGroupEndpoints,'&'));
            $this->agi->setVariable("HG_TIMEOUTLIST", $huntGroup->getRingAllTimeout());

            // use a external number and some internal numbers instead
            $this->agi->setVariable("HG_ENDPOINTLIST", "PJSIP/+34xxxxxxxxx@proxytrunks&PJSIP/aaaaaa_bob&PJSIP/bbbbbb_alice");
```

where xxxxxxxxx is a mobile number and aaaaaa_bob, bbbbbb_alice are internal users.

When I route a DDI to a hunt group with ring all strategy the system correctly calls the internal users and the mobile number.

I understand that this can lead to some problems. If any mobile has a voicemail active or if it is out of coverage area that one is going to answer the call with a locution. But even taking that into account being able to call several external numbers or an external number at the same time than other internal ones can be useful under some circumstances.

For example when the internet connection of a vpbx, retail, residencial goes down so the devices can't be reached we can route the calls to an external number but that is not enough when that DDI receives lots of calls. It would be a lot better to be able to route to several external numbers under this emergency situation.

Besides some people wants their mobile to ring in a hunt group.

For example, I had already configured a pbx where they wanted all their in office VoIP phones in a ring group for 15 seconds and if the call is not answered ring the same all phones and a mobile.

Other one likes all his in office VoIP phones to ring and beside he wants to have the chance to add/remove a mobile number on that huntgroup on special circumstances. The switch is easily done. In ivozprovider would be done with a lock.

Which are reasons why this kind of routing to external numbers on hunt groups is not a good idea?

**Comment by Kaian** [comment]
Hi @manfer 

As you have noticed, external number routing is only available in _leaf_ routes (the ones that has no post-process associated) so we don't make decissions based on their dial statuses. This restriction is made on purpose.

Contrary to UACs talking to Users proxy, external entities handled by Trunks proxy are out of our control and we can not warranty their behaviour. As you mention, calls can answer with a voicemail message (or out-of-network) breaking the huntgroup process. Ring timeouts may also not be accurate when dealing with external entities. It depends a lot on the route taken to call the external device, it make be instant or take a couple of seconds (even consecutive calls to the same destination may have different results).

To sum up, there is no technical limitation to do it, but generally it's not a good idea. 

If you want to call users mobiles and handle them like the terminals, they must be integrated somehow with users proxy and you must have control of how calls to that entities are placed.

Of course this is a free software project and you can change the code to fulfill your needs. Feel free to do your tests as your scenario may be safe enough to ensure the expected behaviour. Glad to see the code can be somehow understood to make changes! 

Regards!

**Comment by Kaian** [comment]
I've been thinking more about this feature and I've actually found another problem for our current agi development status: presentation.

Right now, the caller number is calculated before dialing the destination, depending if the call is to an external number or internal one.

For incoming external calls (i.e. a ddi routed to a Huntgroup) this is not a problem: all callees must see the same origin (the external caller). But for internal calls (i.e. a user calling an extensions that points to the same huntgoup) the presentation may be different for each callee. Users of the huntgoup should see as origin the extension of the caller, while external calls should have caller's outgoingDDI displayed (or even worse, a DDI configured using outgoingDDIRules per destination).

This _different presentation per callee_ is not easy to fix because the origin number must be set in the callers channel (that is just one) for multiple destinations. There are some options to solve this, but none of them convince me yet.

1. Using [pre-dial macros](https://wiki.asterisk.org/wiki/display/AST/Pre-Dial+Handlers)
We use these right now, but just to send some basic headers to the proxies. In a perferct world, this information could be fetched by the proxies and these won't be required, but this way we skip some queries. Adding here presentation logic could be possible (not sure in fact), but this will require more work that one would expect (taking outgoingDDIRules hell into account). The pro's of this development will be that all presentation hell will be taking as the last step of the call and it will be in one logic, instead of spread into multiple code files.

2. Using [party-id macros](https://wiki.asterisk.org/wiki/display/AST/Party+ID+Interception+Macros+and+Routines)
This is a more clean that the previous one, because these ones are designed to update presentation. It would still require the development of the previous point.

3. Using [Local channels](https://wiki.asterisk.org/wiki/display/AST/Local+Channel)
This could be also possible, but less desirable. Instead of calling to final destinations, huntgoup would call to local channels that will invoke an AGI to configure presentation and final destination.

Not sure if you have continued trying to develop this enhancement on your own, but I'll add any other things I came up, just for the record.

Regards!!

**Comment by manfer** [comment]
Hi @Kaian 

I still had not time to look into this in detail but I had taken into account your previous message

> external number routing is only available in leaf routes

And as that would be enough for me, if I try it I will look into creating a new routing target _externals_ that allows to call one or more external numbers and that could be used for example in conditional routes.
